File: hw/vizcaino/u2/hw14MongoDb/hw14MongoDBPy/view/reservation_panel.py
import tkinter as tk
from tkinter import ttk

from dao.reservation_dao import ReservationDAO
from model.reservation import Reservation


class ReservationPanel(tk.Frame):

    def __init__(self, parent):
        super().__init__(parent)

        self.dao = ReservationDAO()

        tk.Label(self, text="ID").grid(row=0, column=0)
        self.txt_id = tk.Entry(self)
        self.txt_id.grid(row=0, column=1)

        tk.Label(self, text="Customer Name").grid(row=1, column=0)
        self.txt_customer_name = tk.Entry(self)
        self.txt_customer_name.grid(row=1, column=1)

        tk.Label(self, text="Phone").grid(row=2, column=0)
        self.txt_phone = tk.Entry(self)
        self.txt_phone.grid(row=2, column=1)

        tk.Label(self, text="Status").grid(row=3, column=0)
        self.txt_status = tk.Entry(self)
        self.txt_status.grid(row=3, column=1)

        self.table = ttk.Treeview(
            self,
            columns=("ID", "Customer Name", "Phone", "Status"),
            show="headings"
        )

        for col in ("ID", "Customer Name", "Phone", "Status"):
            self.table.heading(col, text=col)

        self.table.grid(row=4, column=0, columnspan=4, sticky="nsew")

        tk.Button(self, text="Add", command=self.add_reservation).grid(row=5, column=0)
        tk.Button(self, text="Delete", command=self.delete_reservation).grid(row=5, column=1)
        tk.Button(self, text="Refresh", command=self.load_data).grid(row=5, column=2)

        self.load_data()

    def add_reservation(self):

        reservation = Reservation(
            int(self.txt_id.get()),
            self.txt_customer_name.get(),
            self.txt_phone.get(),
            self.txt_status.get()
        )

        self.dao.save(reservation)

        self.load_data()
        self.clear_fields()

    def delete_reservation(self):

        selected = self.table.selection()

        if selected:

            values = self.table.item(selected[0])["values"]

            self.dao.delete(int(values[0]))

            self.load_data()

    def load_data(self):

        for item in self.table.get_children():
            self.table.delete(item)

        for reservation in self.dao.find_all():

            self.table.insert(
                "",
                tk.END,
                values=(
                    reservation["id"],
                    reservation["customerName"],
                    reservation["phone"],
                    reservation["status"]
                )
            )

    def clear_fields(self):

        self.txt_id.delete(0, tk.END)
        self.txt_customer_name.delete(0, tk.END)
        self.txt_phone.delete(0, tk.END)
        self.txt_status.delete(0, tk.END)

File: hw/vizcaino/u2/hw14MongoDb/hw14MongoDbJava/hw14Interface/src/main/java/ec/edu/espe/view/ReservationPanel.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ec.edu.espe.view;

/**
 *
 * @author Adrian Vizcaino <The-Softwarriors at ESPE>
 */
import ec.edu.espe.model.Reservation;
import ec.edu.espe.model.dao.ReservationDAO;

import java.awt.BorderLayout;
import java.awt.GridLayout;
import java.util.List;

import javax.swing.*;
import javax.swing.table.DefaultTableModel;

public class ReservationPanel extends JPanel {

    private JTextField txtId;
    private JTextField txtCustomerName;
    private JTextField txtPhone;
    private JTextField txtStatus;

    private JTable table;
    private DefaultTableModel model;

    private ReservationDAO dao;

    public ReservationPanel() {

        dao = new ReservationDAO();

        setLayout(new BorderLayout());

        JPanel inputPanel = new JPanel(new GridLayout(4, 2));

        inputPanel.add(new JLabel("ID"));
        txtId = new JTextField();
        inputPanel.add(txtId);

        inputPanel.add(new JLabel("Customer Name"));
        txtCustomerName = new JTextField();
        inputPanel.add(txtCustomerName);

        inputPanel.add(new JLabel("Phone"));
        txtPhone = new JTextField();
        inputPanel.add(txtPhone);

        inputPanel.add(new JLabel("Status"));
        txtStatus = new JTextField();
        inputPanel.add(txtStatus);

        add(inputPanel, BorderLayout.NORTH);

        model = new DefaultTableModel();
        model.addColumn("ID");
        model.addColumn("Customer Name");
        model.addColumn("Phone");
        model.addColumn("Status");

        table = new JTable(model);

        add(new JScrollPane(table), BorderLayout.CENTER);

        JPanel buttonPanel = new JPanel();

        JButton btnAdd = new JButton("Add");
        JButton btnDelete = new JButton("Delete");
        JButton btnRefresh = new JButton("Refresh");

        buttonPanel.add(btnAdd);
        buttonPanel.add(btnDelete);
        buttonPanel.add(btnRefresh);

        add(buttonPanel, BorderLayout.SOUTH);

        btnAdd.addActionListener(e -> addReservation());
        btnDelete.addActionListener(e -> deleteReservation());
        btnRefresh.addActionListener(e -> loadData());

        loadData();
    }

    private void addReservation() {

        try {

            Reservation reservation = new Reservation(
                    Integer.parseInt(txtId.getText()),
                    txtCustomerName.getText(),
                    txtPhone.getText(),
                    txtStatus.getText()
            );

            dao.save(reservation);
            loadData();
            clearFields();

        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    }

    private void deleteReservation() {

        int row = table.getSelectedRow();

        if (row >= 0) {

            int id = Integer.parseInt(model.getValueAt(row, 0).toString());

            dao.delete(id);
            loadData();
        }
    }

    private void loadData() {

        model.setRowCount(0);

        List<Reservation> reservations = dao.findAll();

        for (Reservation r : reservations) {

            model.addRow(new Object[]{
                    r.getId(),
                    r.getCustomerName(),
                    r.getPhone(),
                    r.getStatus()
            });
        }
    }

    private void clearFields() {

        txtId.setText("");
        txtCustomerName.setText("");
        txtPhone.setText("");
        txtStatus.setText("");
    }
}